doc/ml_features.rst

```diff
@@ -0,0 +1,486 @@
+Instead of Neural Networks, or, in general, Machine Learning models, it is
+possible to upload to the DB (collections of) functions which can be used e.g.
+to perform pre- or post-processing operations on tensors stored on the DB.
+
+Since the functions are going to be stored as TorchScript modules, they
+
+- need to be jit-traceable
+- can use ``torch`` as a built-in module
+- can **not** import modules
+
+In this section we will see how to
+
+- save a collection of functions to a script file, upload them to the DB,
+  and execute them on tensors stored on the DB.
+- define and upload a function on-the-fly from a Python script and
+  execute it on tensors stored on the DB.
+
+
+=================================================================
+Uploading a script containing a collection of functions to the DB
+=================================================================
+
+The easiest way of defining and storing functions on the DB is to create a
+dedicated file. In that file, we can define functions which will be callable
+through the SmartRedis client, but also from other functions in the
+same file. A typical script file would look like this:
+
+.. code-block:: python
+
+    def rescale(tensor, mu: float, sigma: float):
+        mean = tensor.mean()
+        std = tensor.std()
+
+        normalized = (tensor-mean)/std
+        return tensor*sigma + mu
+
+    def shift_y_to_x(x, y):
+        mu_x = x.mean()
+        sigma_x = x.std()
+        y_rescaled = rescale(y, mu_x, sigma_x)
+
+        return y_rescaled
+
+In the script, we defined ``shift_y_to_x``,
+a function which returns a modified copy of a tensor ``y``,
+which matches the statistical distribution of the tensor ``x``.
+Notice that we are not importing ``torch`` in the script, as it will
+be recognized as a built-in by the TorchScript compiler. Because
+of the discrepancy between TorchScript's and Python's syntaxes, TorchScript
+scripts cannot be run as standalone Python scripts.
+
+Here is the code which allows us to run the function ``shift_y_to_x`` on
+tensors stored in the DB. We will assume that the above script is stored
+as ``"./shift.script"``.
+
+.. code-block:: python
+
+    import numpy as np
+    from smartredis import Client
+
+    # Generate tensors according to two different random distributions
+    x = np.random.rand(100, 100).astype(np.float32)
+    y = np.random.rand(100, 100).astype(np.float32) * 2 + 10
+
+    # Instantiate and connect SmartRedis client
+    client = Client(cluster=False)
+
+    # Upload tensors to DB
+    client.put_tensor("X_rand", x)
+    client.put_tensor("Y_rand", y)
+
+    # Upload script containing functions to DB
+    client.set_script_from_file("shifter", "./shift.script", device="CPU")
+    # Run the function ``shift_y_to_x`` on ``X_rand`` and ``Y_rand``
+    client.run_script("shifter", "shift_y_to_x", inputs=["X_rand", "Y_rand"], outputs=["Y_scaled"])
+    # Download output
+    y_scaled = client.get_tensor("Y_scaled")
+
```

Comment:
I would add a mini end summary here, something like Notice how I use the function ``Client.set_script_from_file()`` to upload the script to the DB, then call the script from within the DB via ``Client.run_script()``, the output tensors are stored in the DB so I use get_tensor to retrieve

Reply:
ermmm I see you actually have this in the code comments, maybe its actually fine

Reply:
I am able to understand it and follow by reading the comments btw!

Reply:
Added a small recap anyhow!